### PR TITLE
Console logger support for IncludeEvaluationPropertiesAndItems

### PR DESCRIPTION
**Pull request by KirillOsenkov**
This PR supersedes https://github.com/dotnet/msbuild/pull/6514

Console logger support for IncludeEvaluationPropertiesAndItems.

Includes https://github.com/dotnet/msbuild/pull/6520, but we can take this commit out if it's merged separately.

OutputItems was pretty much duplicated in ParallelConsoleLogger. Unify with the base implementation and extract methods that need to be replaced by the derived type.

Support for reading project configuration description either from ProjectStartedEventArgs.Items or ProjectEvaluationFinishedEventArgs.Items.

Skip TestItemsWithUnexpandableMetadata. Issue https://github.com/dotnet/msbuild/issues/6518 is tracking. More work is needed to understand the desired behavior of the system under test and then fix the test to test the desired behavior.

Skip NullMetadataOnLegacyOutputItems_InlineTask. The feature the test is testing is broken, but the test passes because it doesn't specify diag verbosity for its logger. We will only log task outputs with diag verbosity. Issue https://github.com/dotnet/msbuild/issues/6521 is tracking.

Opt test out of LogPropertiesAndItemsAfterEvaluation. OutOfProcNodeForwardCertainproperties is explicitly testing a feature where only some properties are forwarded from a different process on ProjectStartedEventArgs. When we move properties to ProjectEvaluationFinished the test loses meaning. So force it into the old behavior via an escape hatch.

### CODE REVIEW

File: src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs [hunk 1]
```diff
@@ -540,20 +540,38 @@ public override void ProjectStartedHandler(object sender, ProjectStartedEventArg
                 }
             }
 
-            ReadProjectConfigurationDescription(e.BuildEventContext, e.Items);
+            var projectKey = (e.BuildEventContext.NodeId, e.BuildEventContext.ProjectContextId);
```
Comment (Forgind) [inline src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs:543]: Can e.BuildEventContext be null here?

**Review by Forgind — COMMENTED** [review]
I think this should work, though some of the unification with BaseConsoleLogger was hard to follow. Most worried about the possible NRE.

File: src/Build/BackEnd/Components/Logging/LoggingService.cs [hunk 2]
```diff
@@ -514,7 +514,18 @@ public bool IncludeTaskInputs
...
+            get
+            {
+                if (_includeEvaluationPropertiesAndItems == null)
+                {
+                    var sinks = _eventSinkDictionary.Values.OfType<EventSourceSink>();
+                    // .All() on an empty list defaults to true, we want to default to false
+                    _includeEvaluationPropertiesAndItems = sinks.Any() && sinks.All(sink => sink.IncludeEvaluationPropertiesAndItems);
+                }
+
+                return _includeEvaluationPropertiesAndItems ?? false;
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:526]: sorry what? the nullable? no, it will be assigned to a boolean on line 523, but that doesn't convince the compiler, so I had to add `?? false` to convert the `bool?` to `bool`

File: src/Build/Logging/BaseConsoleLogger.cs [hunk 3]
```diff
@@ -642,42 +644,56 @@ internal SortedList ExtractItemList(IEnumerable items)
         /// </summary>
         internal virtual void OutputItems(string itemType, ArrayList itemTypeList)
         {
-            // Write each item, one per line
-            bool haveWrittenItemType = false;
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BaseConsoleLogger.cs:646]: yes, it was completely unnecessary to do it inside the loop

File: src/Build/Logging/BaseConsoleLogger.cs [hunk 4]
```diff
@@ -642,42 +644,56 @@ internal SortedList ExtractItemList(IEnumerable items)
...
+                var metadata = item switch
                 {
-                    string valueOrError;
-                    try
-                    {
-                        valueOrError = item.GetMetadata(metadatum.Key as string);
-                    }
-                    catch (InvalidProjectFileException e)
+                    IMetadataContainer metadataContainer => metadataContainer.EnumerateMetadata(),
+                    IItem<ProjectMetadata> iitem => iitem.Metadata?.Select(m => new KeyValuePair<string, string>(m.Name, m.EvaluatedValue)),
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BaseConsoleLogger.cs:664]: not sure about inlining

good question about unescaping, need to see whether the values on iitem.Metadata are already unescaped...

File: src/Build/Logging/BaseConsoleLogger.cs [hunk 5]
```diff
@@ -642,42 +644,56 @@ internal SortedList ExtractItemList(IEnumerable items)
...
+        protected virtual void WriteItemType(string itemType)
+        {
+            setColor(ConsoleColor.Gray);
+            WriteLinePretty(itemType);
+            setColor(ConsoleColor.DarkGray);
+        }
+
+        protected virtual void WriteItemSpec(string itemSpec)
+        {
+            WriteLinePretty("    " + itemSpec);
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BaseConsoleLogger.cs:689]: it won't allocate the string literal, but it will allocate the string for the concatenated total, and it's terrible.

Fortunately we may have an intern this summer who may help look into rewriting the console loggers to avoid allocations.

File: src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs (hunk 1 above)
Comment (KirillOsenkov) [reply src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs:543]: No, BuildEventContext should always be set

**Review by Forgind — APPROVED** [review]

File: src/Build/BackEnd/Components/Logging/LoggingService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:526]: No, I don't think the bang will work to convert bool? to bool:
https://sharplab.io/#v2:C4LghgzsA0AmIGoA+ABATARgLACgUGYACdQgYUIG9dCbiiUAWQgWQAoBKS62ngIwHt+AGwD8hXoQC8hAHYBXIUIDc3HjQHDxaKeMIixAMzBCIAUxU41tDUPFFpvAIQWeAX1yugA=

[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:526]: <confusion>
I thought that was exactly what the bang was supposed to do. Ah, well.

[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:526]: ah, I see :) The bang is used to force Nullable Reference Types (the new hotness), but ?? x is used to convert a nullable Value Type to the underlying value type. Nullable value types (Nullable<T>) was introduced way back in C# 2 in 2005

**Review by rainersigwald — APPROVED** [review]
Looks reasonable to me.

On the larger issue, I have a thought; I bet you can explain why it's bad. Could we log the data in both places, but only have the binlog record it at end-of-evaluation? And when replaying a binlog read it from the correlated end-of-evaluation to the ProjectStarted?

**Comment by KirillOsenkov** [comment]
It's actually a reasonable thought. I don't see why it would be bad, perhaps it's worth implementing.

Remember though you'll only get the properties and items on ProjectStarted in the central node, since they are not serialized across. But it would achieve backwards compatibility (parity with what it used to be). Perhaps I should have built it this way to begin with!

The other question is combined with the other workarounds we've built, is it necessary anymore?

**Comment by KirillOsenkov** [comment]
One thing that comes to mind is populating properties and items on ProjectStarted would decrease performance and increase allocations (since we would be walking the expensive data structures and copy-on-writing). So we'd lose some of those hard-won perf wins that we so tightly optimized.
